_posts/2021-03-15-forecasting-time-series-with-torch_3/forecasting_3.Rmd

```diff
@@ -0,0 +1,445 @@
+
+`state`, too, is polyvalent. But here, there are just two possibilities.
+
+-   The first time the decoder is called, it is "seeded" with the final state from the encoder. Note how this is *the only time* we make use of the encoding.
+
+-   From then on, the decoder's own previous state will be passed. Remember how it returns two values, forecast and state?
+
+```{r}
+seq2seq_module <- nn_module(
+  
+  initialize = function(type, input_size, hidden_size, n_forecast, num_layers = 1, encoder_dropout = 0) {
+    
+    self$encoder <- encoder_module(type = type, input_size = input_size,
+                                   hidden_size = hidden_size, num_layers, encoder_dropout)
+    self$decoder <- decoder_module(type = type, input_size = input_size,
+                                   hidden_size = hidden_size, num_layers)
+    self$n_forecast <- n_forecast
+    
+  },
+  
+  forward = function(x, y, teacher_forcing_ratio) {
+    
+    # prepare empty output
+    outputs <- torch_zeros(dim(x)[1], self$n_forecast)$to(device = device)
+    
+    # encode current input sequence
+    hidden <- self$encoder(x)
+    
+    # prime decoder with final input value and hidden state from the encoder
+    out <- self$decoder(x[ , n_timesteps, , drop = FALSE], hidden)
+    
+    # decompose into predictions and decoder state
+    # pred is (batch_size, 1)
+    # state is (1, batch_size, hidden_size)
+    c(pred, state) %<-% out
+    
+    # store first prediction
+    outputs[ , 1] <- pred$squeeze(2)
+    
+    # iterate to generate remaining forecasts
+    for (t in 2:self$n_forecast) {
+      
+      # call decoder on either ground truth or previous prediction, plus previous decoder state
+      teacher_forcing <- runif(1) < teacher_forcing_ratio
+      input <- if (teacher_forcing == TRUE) y[ , t - 1, drop = FALSE] else pred
+      input <- input$unsqueeze(3)
+      out <- self$decoder(input, state)
+      
+      # again, decompose decoder return values
+      c(pred, state) %<-% out
+      # and store current prediction
+      outputs[ , t] <- pred$squeeze(2)
+    }
+    outputs
+  }
+  
+)
+
+net <- seq2seq_module("gru", input_size = 1, hidden_size = 32, n_forecast = n_forecast)
+
+# training RNNs on the GPU currently prints a warning that may clutter 
+# the console
+# see https://github.com/mlverse/torch/issues/461
+# alternatively, use 
+# device <- "cpu"
+device <- torch_device(if (cuda_is_available()) "cuda" else "cpu")
+
+net <- net$to(device = device)
+```
+
+# Training
+
+The training procedure is *mainly* unchanged. We do, however, need to decide about `teacher_forcing_ratio`, the proportion of input sequences we want to perform re-calibration on. In `valid_batch`, this should always be `0`, while in `train_batch`, it's up to us (or rather, experimentation). Here, we set it to `0.3`.
```

Comment:
```suggestion
The training procedure is *mainly* unchanged. We do, however, need to decide about `teacher_forcing_ratio`, the proportion of input sequences we want to perform re-calibration on. In `valid_batch()`, this should always be `0`, while in `train_batch()`, it's up to us (or rather, experimentation). Here, we set it to `0.3`.
```

Reply:
Functions get trailing parentheses to distinguish them from arguments.

Reply:
forgot, sorry